### ISSUE TITLE
 Implement `TryFrom<&OsStr>` for `&str`

### ISSUE THREAD

**Issue by aticu** [rust-lang/libs-team#53]
# Proposal

## Problem statement

Currently it is impossible through use of the standard conversion traits to convert from an `&OsStr` to a `&str`.

## Motivation, use-cases

Adding an implementation of `TryFrom<&OsStr>` for `&str` makes this conversion more discoverable to users who are already familiar with the `TryFrom` trait. They may even expect such an implementation to exist.
Additionally it allows the use of `&OsStr` in more generic methods such as those that  accept `impl TryInto<&str>`:

```rust
fn maybe_print<'a>(arg: impl TryInto<&'a str>) {
    if let Ok(converted) = arg.try_into() {
        println!("{converted}");
    }
}
```

## Solution sketches

Implement `TryFrom<&OsStr>` for `&str`.

## Links and related work

Implementation PR: rust-lang/rust#98202

## What happens now?

This issue is part of the libs-api team [API change proposal process]. Once this issue is filed the libs-api team will review open proposals in its weekly meeting. You should receive feedback within a week or two.

[API change proposal process]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html

**Comment by yaahc** [comment]
@aticu can you give a more concrete example of a situation where you'd use this?

**Comment by aticu** [comment]
I don't remember the exact details (and I currently don't have access to the code), but as far as I remember I was trying to do some checks on the contents of an `&OsStr` and if they matched some predicates I wanted to further process them.
Since there is no direct way to look at the contents of an `&OsStr`, I was fine with just doing these checks for those `&OsStr` that are also valid `&str`s.
So naturally I tried `os_str.try_into()` and was surprised to learn that this implementation does not exist. Then when looking at the implementation I found the `to_str` method of `&OsStr`, which also worked in my usecase.

So for my specific situation this `impl` wouldn't have made anything possible that isn't already possible, but it would have saved me from confusion why the trait isn't implemented (as at least to me it seems like it should be) and it would have saved me some time looking at documentation.
With that reasoning I thought it might also be helpful for others and thus decided to try add that `impl`.

**Comment by thomcc** [comment]
I think it's reasonable for this conversion to exist -- it's a fallible conversion that exists as an inherent method, it seems inconsistent if we're going to decide that it shouldn't exist via our trait for fallible conversions.

Or am I missing something that makes this case less straightforward than it seems? (perhaps we'd change things in a world where we guarantee the encoding?)

**Comment by yaahc** [comment]
> I think it's reasonable for this conversion to exist -- it's a fallible conversion that exists as an inherent method, it seems inconsistent if we're going to decide that it shouldn't exist via our trait for fallible conversions.
> 
> Or am I missing something that makes this case less straightforward than it seems? (perhaps we'd change things in a world where we guarantee the encoding?)

I agree and I don't think you're missing anything, I just wanted the concrete use case because I think its often very useful context. Also, I just realized this change isn't unstable, it's insta-stable, so it will need an FCP not an ACP.

**Comment by yaahc** [comment]
I'm retagging this since it's an insta-stable change that needs an FCP.

Along those lines, I'm not sure if rfcbot is setup to work on this repo, but there's an easy way to find out.

@rfcbot merge
